pkg/controller/oam/applicationconfiguration/applicationconfiguration.go

```diff
@@ -0,0 +1,311 @@
+			trait:    ResourceRenderFn(renderTrait),
+		},
+		workloads: &workloads{
+			client: m.GetClient(),
+		},
+		gc:     GarbageCollectorFn(eligible),
+		log:    logging.NewNopLogger(),
+		record: event.NewNopRecorder(),
+	}
+
+	for _, ro := range o {
+		ro(r)
+	}
+
+	return r
+}
+
+// NOTE(negz): We don't validate anything against their definitions at the
+// controller level. We assume this will be done by validating admission
+// webhooks.
+
+// Reconcile an OAM ApplicationConfigurations by rendering and instantiating its
+// Components and Traits.
+func (r *Reconciler) Reconcile(req reconcile.Request) (reconcile.Result, error) {
+	log := r.log.WithValues("request", req)
+	log.Debug("Reconciling")
+
+	ctx, cancel := context.WithTimeout(context.Background(), reconcileTimeout)
+	defer cancel()
+
+	ac := &v1alpha2.ApplicationConfiguration{}
+	if err := r.client.Get(ctx, req.NamespacedName, ac); err != nil {
+		return reconcile.Result{}, errors.Wrap(resource.IgnoreNotFound(err), errGetAppConfig)
+	}
+
+	log = log.WithValues("uid", ac.GetUID(), "version", ac.GetResourceVersion())
+
+	workloads, err := r.components.Render(ctx, ac)
+	if err != nil {
+		log.Debug("Cannot render components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotRenderComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errRenderComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully rendered components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonRenderComponents, "Successfully rendered components", "workloads", strconv.Itoa(len(workloads))))
+
+	if err := r.workloads.Apply(ctx, workloads); err != nil {
+		log.Debug("Cannot apply components", "error", err, "requeue-after", time.Now().Add(shortWait))
+		r.record.Event(ac, event.Warning(reasonCannotApplyComponents, err))
+		ac.Status.SetConditions(v1alpha1.ReconcileError(errors.Wrap(err, errApplyComponents)))
+		return reconcile.Result{RequeueAfter: shortWait}, errors.Wrap(r.client.Status().Update(ctx, ac), errUpdateAppConfigStatus)
+	}
+	log.Debug("Successfully applied components", "workloads", len(workloads))
+	r.record.Event(ac, event.Normal(reasonApplyComponents, "Successfully applied components", "workloads", strconv.Itoa(len(workloads))))
+
+	for _, e := range r.gc.Eligible(ac.GetNamespace(), ac.Status.Workloads, workloads) {
+		// https://github.com/golang/go/wiki/CommonMistakes#using-reference-to-loop-iterator-variable
+		e := e
+
+		log := log.WithValues("kind", e.GetKind(), "name", e.GetName())
+		record := r.record.WithAnnotations("kind", e.GetKind(), "name", e.GetName())
+
+		if err := r.client.Delete(ctx, &e); resource.IgnoreNotFound(err) != nil {
+			log.Debug("Cannot garbage collect component", "error", err, "requeue-after", time.Now().Add(shortWait))
```

Comment:
Nit: I think we are saying `"requeue-after"` here, but then providing the actual expected time of requeue rather than how long until requeue. This is present in a few other places, but I will not denote all as to not clutter this PR with more comments. Let me know if I am mistaken here :)

Reply:
That's intended. I felt like it was better for the logs to print `"requeue-after": "5:00pm"` rather than `"requeue-after": "30 seconds"`, which would require the reader to add the duration to the log timestamp to figure out when the next reconcile would happen.